File: PageObjects/HomePage.cs
using Microsoft.Playwright;

namespace PerfectDraftTests.PageObjects
{
    public class HomePage : BasePage
    {
        private readonly IPage _page;

        public HomePage(IPage page) : base(page)
        {
            _page = page;
        }

        public async Task NavigateToUKWebsite()
        {
            await _page.GotoAsync("https://www.perfectdraft.com/en-gb");
            await _page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
        }

        public async Task<bool> IsNavigationMenuVisible()
        {
            // Look for main navigation - trying multiple selectors
            var navSelectors = new[] { "nav", ".navigation", ".main-nav", "[role='navigation']" };

            foreach (var selector in navSelectors)
            {
                try
                {
                    var element = await _page.QuerySelectorAsync(selector);
                    if (element != null && await element.IsVisibleAsync())
                    {
                        return true;
                    }
                }
                catch
                {
                    continue;
                }
            }
            return false;
        }

        public async Task<bool> IsMenuItemVisible(string menuItem)
        {
            // Use Playwright's text selector which is more reliable
            try
            {
                var element = await _page.QuerySelectorAsync($"text={menuItem}");
                return element != null && await element.IsVisibleAsync();
            }
            catch
            {
                return false;
            }
        }

        public async Task<bool> IsSearchFunctionalityAvailable()
        {
            var searchSelectors = new[]
            {
                "input[type='search']",
                "[placeholder*='Search']",
                "[placeholder*='search']",
                "input[name*='search']",
                ".search input"
            };

            foreach (var selector in searchSelectors)
            {
                try
                {
                    var element = await _page.QuerySelectorAsync(selector);
                    if (element != null && await element.IsVisibleAsync())
                    {
                        return true;
                    }
                }
                catch
                {
                    continue;
                }
            }
            return false;
        }

        public async Task<string> GetCartItemCount()
        {
            var cartSelectors = new[]
            {
                ".cart-count",
                ".cart-counter",
                "[data-cart-count]",
                ".basket-count"
            };

            foreach (var selector in cartSelectors)
            {
                try
                {
                    var element = await _page.QuerySelectorAsync(selector);
                    if (element != null && await element.IsVisibleAsync())
                    {
                        var text = await element.TextContentAsync();
                        return text?.Trim() ?? "0";
                    }
                }
                catch
                {
                    continue;
                }
            }

            // If no specific counter found, default to "0"
            return "0";
        }

        public async Task<bool> IsCartIconVisible()
        {
            var cartSelectors = new[]
            {
                "text=Cart",
                "text=Basket",
                "[href*='cart']",
                "[href*='basket']",
                ".cart-icon",
                ".basket-icon"
            };

            foreach (var selector in cartSelectors)
            {
                try
                {
                    var element = await _page.QuerySelectorAsync(selector);
                    if (element != null && await element.IsVisibleAsync())
                    {
                        return true;
                    }
                }
                catch
                {
                    continue;
                }
            }
            return false;
        }

        public async Task NavigateToSection(string sectionName)
        {
            if (sectionName == "Kegs")
            {
                try
                {
                    // First try direct navigation which is more reliable
                    await _page.GotoAsync("https://www.perfectdraft.com/en-gb/perfect-draft-range/perfect-draft-kegs");
                    await _page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
                    Console.WriteLine("Successfully navigated directly to kegs page");
                }
                catch
                {
                    // Fallback to clicking the navigation link
                    try
                    {
                        // Use the specific main navigation link for kegs with force click
                        var kegsLink = _page.Locator("a[data-menu='menu-605']").First;
                        await kegsLink.ClickAsync(new() { Force = true });
                    }
                    catch
                    {
                        // Try alternative navigation approaches
                        await _page.EvaluateAsync("window.location.href = 'https://www.perfectdraft.com/en-gb/perfect-draft-range/perfect-draft-kegs'");
                    }
                }
            }
            else if (sectionName == "Machines" || sectionName == "PerfectDraft Machines")
            {
                try
                {
                    // Use the specific link for machines
                    var machinesLink = _page.GetByText("PerfectDraft Machines").First;
                    await machinesLink.ClickAsync(new() { Force = true });
                }
                catch
                {
                    // Direct navigation fallback
                    await _page.GotoAsync("https://www.perfectdraft.com/en-gb/perfect-draft-range/perfect-draft-machines");
                }
            }
            else
            {
                // Fallback to text search for other sections
                try
                {
                    var link = _page.GetByText(sectionName).First;
                    await link.ClickAsync(new() { Force = true });
                }
                catch
                {
                    Console.WriteLine($"Could not navigate to section: {sectionName}");
                }
            }
            
            await _page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
            await _page.WaitForTimeoutAsync(2000); // Allow time for dynamic content to load
        }

        public async Task<bool> IsFooterLinkVisible(string linkText)
        {
            // Common footer selectors
            var footerSelectors = new[]
            {
                "footer",
                ".footer",
                "[role='contentinfo']",
                ".site-footer",
                "#footer"
            };

            // First try to find the footer
            foreach (var footerSelector in footerSelectors)
            {
                try
                {
                    var footer = await _page.QuerySelectorAsync(footerSelector);
                    if (footer != null && await footer.IsVisibleAsync())
                    {
                        // Look for the link within the footer
                        var linkInFooter = await footer.QuerySelectorAsync($"text={linkText}");
                        if (linkInFooter != null && await linkInFooter.IsVisibleAsync())
                        {
                            return true;
                        }
                        
                        // Also try with a partial match
                        var partialLinkInFooter = await footer.QuerySelectorAsync($"text=/{linkText}/i");
                        if (partialLinkInFooter != null && await partialLinkInFooter.IsVisibleAsync())
                        {
                            return true;
                        }
                    }
                }
                catch
                {
                    continue;
                }
            }

            // Fallback: try to find the link anywhere on the page
            try
            {
                var element = await _page.QuerySelectorAsync($"text={linkText}");
                return element != null && await element.IsVisibleAsync();
            }
            catch
            {
                return false;
            }
        }

        public async Task ClickFooterLink(string linkText)
        {
            // Common footer selectors
            var footerSelectors = new[]
            {
                "footer",
                ".footer",
                "[role='contentinfo']",
                ".site-footer",
                "#footer"
            };

            // First try to find and click the link within the footer
            foreach (var footerSelector in footerSelectors)
            {
                try
                {
                    var footer = await _page.QuerySelectorAsync(footerSelector);
                    if (footer != null && await footer.IsVisibleAsync())
                    {
                        // Look for the exact text link within the footer
                        var linkInFooter = await footer.QuerySelectorAsync($"text={linkText}");
                        if (linkInFooter != null && await linkInFooter.IsVisibleAsync())
                        {
                            await linkInFooter.ClickAsync();
                            await _page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
                            return;
                        }
                        
                        // Try with partial match
                        var partialLinkInFooter = await footer.QuerySelectorAsync($"text=/{linkText}/i");
                        if (partialLinkInFooter != null && await partialLinkInFooter.IsVisibleAsync())
                        {
                            await partialLinkInFooter.ClickAsync();
                            await _page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
                            return;
                        }
                    }
                }
                catch
                {
                    continue;
                }
            }

            // Fallback: try to find and click the link anywhere on the page
            try
            {
                var element = await _page.QuerySelectorAsync($"text={linkText}");
                if (element != null && await element.IsVisibleAsync())
                {
                    await element.ClickAsync();
                    await _page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
                    return;
                }

                // Try with case-insensitive partial match
                var partialElement = await _page.QuerySelectorAsync($"text=/{linkText}/i");
                if (partialElement != null && await partialElement.IsVisibleAsync())
                {
                    await partialElement.ClickAsync();
                    await _page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
                    return;
                }
            }
            catch
            {
                // If all else fails, try direct navigation
                if (linkText.ToLower().Contains("about"))
                {
                    await _page.GotoAsync("https://www.perfectdraft.com/en-gb/about-perfectdraft");
                    await _page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
                }
            }
        }
    }
}

File: src/pages/home.page.js
// src/pages/home.page.js
import BasePage from './base.page.js';

/**
 * Home Page Object
 * Represents the PerfectDraft home page
 */
class HomePage extends BasePage {
  constructor(page) {
    super(page);

    // Single specific selectors
    this.navigationSelector = 'nav.main-nav';
    this.searchSelector = 'input[type="search"]';
    this.cartCountSelector = '.counter-number';
    this.cartIconSelector = '.minicart-wrapper > a.showcart';
  }

  /**
   * Navigate to UK website
   */
  async navigateToUKWebsite() {
    await this.page.goto('https://www.perfectdraft.com/en-gb', { waitUntil: 'domcontentloaded' });
  }

  /**
   * Navigate to country-specific website
   * @param {string} countryCode - Country code (e.g., 'GB', 'FR')
   */
  async navigateToCountry(countryCode) {
    const url = this.config.getCountryUrl(countryCode);
    await this.page.goto(url, { waitUntil: 'domcontentloaded' });
  }

  /**
   * Check if navigation menu is visible
   * @returns {Promise<boolean>}
   */
  async isNavigationMenuVisible() {
    try {
      const element = this.page.locator(this.navigationSelector);
      return await element.isVisible({ timeout: 5000 });
    } catch {
      return false;
    }
  }

  /**
   * Check if menu item is visible
   * @param {string} menuItem - Menu item text
   * @returns {Promise<boolean>}
   */
  async isMenuItemVisible(menuItem) {
    try {
      const element = await this.page.$(`text=${menuItem}`);
      return element !== null && await element.isVisible();
    } catch {
      return false;
    }
  }

  /**
   * Check if search functionality is available
   * @returns {Promise<boolean>}
   */
  async isSearchFunctionalityAvailable() {
    // Wait for page to be fully loaded since Algolia search loads dynamically
    await this.page.waitForLoadState('networkidle', { timeout: 10000 });

    try {
      const element = this.page.locator(this.searchSelector);
      return await element.isVisible({ timeout: 5000 });
    } catch {
      return false;
    }
  }

  /**
   * Get cart item count
   * @returns {Promise<string>}
   */
  async getCartItemCount() {
    try {
      const element = this.page.locator(this.cartCountSelector);
      await element.waitFor({ timeout: 5000 });
      const text = await element.textContent();
      return text?.trim() || '0';
    } catch {
      return '0';
    }
  }

  /**
   * Check if cart icon is visible
   * @returns {Promise<boolean>}
   */
  async isCartIconVisible() {
    try {
      const element = this.page.locator(this.cartIconSelector);
      return await element.isVisible({ timeout: 5000 });
    } catch {
      return false;
    }
  }

  /**
   * Navigate to section
   * @param {string} sectionName - Section name
   */
  async navigateToSection(sectionName) {
    if (sectionName === 'Kegs') {
      try {
        // First try direct navigation which is more reliable
        await this.page.goto('https://www.perfectdraft.com/en-gb/perfect-draft-range/perfect-draft-kegs');
        await this.page.waitForLoadState('domcontentloaded');
        console.log('Successfully navigated directly to kegs page');
      } catch {
        // Fallback to clicking the navigation link
        try {
          // Use the specific main navigation link for kegs with force click
          const kegsLink = this.page.locator('a[data-menu="menu-605"]').first();
          await kegsLink.click({ force: true });
        } catch {
          // Try alternative navigation approaches
          await this.page.evaluate(() => {
            window.location.href = 'https://www.perfectdraft.com/en-gb/perfect-draft-range/perfect-draft-kegs';
          });
        }
      }
    } else if (sectionName === 'Machines' || sectionName === 'PerfectDraft Machines') {
      // Use more specific selector targeting anchor elements only
      const machinesLink = this.page.locator('a').getByText('PerfectDraft Machines');
      await machinesLink.click({ force: true });

      // Wait for navigation to complete - bundles page specific content
      await this.page.locator('.ais-InfiniteHits').waitFor();
    } else {
      // Fallback to text search for other sections
      try {
        const link = this.page.getByText(sectionName).first();
        await link.click({ force: true });
      } catch {
        console.log(`Could not navigate to section: ${sectionName}`);
      }
    }

    await this.page.waitForLoadState('domcontentloaded');
    await this.page.waitForTimeout(2000); // Allow time for dynamic content to load
  }

  /**
   * Click on login/account link
   */
  async clickLoginLink() {
    // Use the specific login selector for PerfectDraft
    const loginLink = this.page.locator('a[href*="/customer/account"]').first();
    
    if (await loginLink.isVisible({ timeout: 5000 })) {
      await loginLink.click();
    } else {
      throw new Error('Could not find login link');
    }
  }

  /**
   * Accept cookies if banner is present
   */
  async acceptCookies() {
    // Use the specific OneTrust cookie consent button
    const cookieButton = this.page.locator('#onetrust-accept-btn-handler');
    
    if (await cookieButton.isVisible({ timeout: 3000 })) {
      await cookieButton.click();
      await this.page.waitForTimeout(500); // Wait for banner to disappear
      return true;
    }
    return false;
  }

  /**
   * Search for a product
   * @param {string} searchTerm - Search term
   */
  async searchForProduct(searchTerm) {
    const searchInput = this.page.locator(this.searchSelector);
    await searchInput.waitFor({ timeout: 5000 });

    await searchInput.fill(searchTerm);
    await searchInput.press('Enter');

    await this.page.waitForLoadState('networkidle');
  }

  /**
   * Get featured products
   * @returns {Promise<Array>}
   */
  async getFeaturedProducts() {
    // Use specific selector for product cards on PerfectDraft homepage
    const products = await this.page.$$('.product-item');
    
    if (products.length > 0) {
      const productData = [];
      for (const product of products) {
        try {
          const name = await product.$eval('.product-item-name', el => el.textContent?.trim());
          const price = await product.$eval('.price', el => el.textContent?.trim());
          productData.push({ name, price });
        } catch {
          // Skip this product if selectors don't match
          continue;
        }
      }
      return productData;
    }
    return [];
  }

  /**
   * Check if promotional banner is visible
   * @returns {Promise<boolean>}
   */
  async isPromotionalBannerVisible() {
    try {
      const element = this.page.locator('.hero-banner');
      return await element.isVisible({ timeout: 5000 });
    } catch {
      return false;
    }
  }

  /**
   * Get page content for verification
   * @returns {Promise<string>}
   */
  async getPageContent() {
    return await this.page.textContent('body') || '';
  }

  /**
   * Get HTML language attribute
   * @returns {Promise<string|null>}
   */
  async getHtmlLang() {
    return await this.page.getAttribute('html', 'lang');
  }

  /**
   * Wait for country-specific URL
   * @param {string} urlPattern - URL pattern to wait for
   */
  async waitForCountryUrl(urlPattern) {
    await this.page.waitForURL(new RegExp(urlPattern));
  }
}

export default HomePage;
